src/bin/pg_autoctl/service_monitor_init.c

```diff
@@ -146,6 +146,17 @@ service_monitor_init_start(void *context, pid_t *pid)
 			}
 			else
 			{
+				/*
+				 * We are in a sub-process and didn't call exec() on our
+				 * pg_autoctl do service listener program yet we do not want to
+				 * clean-up the semaphore just yet. Publish that we are a
+				 * sub-process and only then quit, avoiding to call the
+				 * atexit() semaphore clean-up function.
+				 */
+				IntString semIdString = intToString(log_semaphore.semId);
+
+				setenv(PG_AUTOCTL_LOG_SEMAPHORE, semIdString.strValue, 1);
+
```

Comment:
I think it makes sense to do this right away after every `fork` that returns `0` in the codebase.

Reply:
That's another way to do it, but when `createAndRun` is true we are going to call `service_monitor_runprogram` which takes care of it anyway. And doing it in the child arm of the fork begs the question of always doing that in other fork call sites, and I don't think we should.

Sending a code revision following your comments so that we can see if we like it better that way or not.

Reply:
> And doing it in the child arm of the fork begs the question of always doing that in other fork call sites, and I don't think we should.

Why should we not do that? It seems to me that only the parent process should kill the semaphore `atexit`, so doing it after every `fork` call that indicates it's a child seems a good thing.

Maybe it would be clearer to set a global indicating if it is the parent process or not, and only do the semaphore unlinking in that case.

Reply:
The problem we have here is that the signals and globals and many things are shared at `fork()` time and lost at `exec()` time. The `atexit` registrations have that policy, so that when we do our run program (which calls `exec`) we don't have the `atexit` functions anymore. 

There's only one place where we `fork()` and then we don't `exec()`, it's when doing `pg_autoctl create monitor` without the `--run` option, because we don't `exec()` the monitor installer.